File: cursor/scripts/save-token-usage.py
#!/usr/bin/env python3
"""Append a token usage snapshot to local JSONL history."""

from __future__ import annotations

import argparse
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


DEFAULT_HISTORY = Path("~/.cursor/token-tracker/history.jsonl").expanduser()
INT_FIELDS = ("prompt_tokens", "completion_tokens", "total_tokens")


def fail(message: str) -> None:
    print(f"token-tracker: {message}", file=sys.stderr)
    raise SystemExit(2)


def load_payload(args: argparse.Namespace) -> dict[str, Any]:
    raw = args.json
    if raw is None and not sys.stdin.isatty():
        raw = sys.stdin.read()
    if raw:
        try:
            payload = json.loads(raw)
        except json.JSONDecodeError as exc:
            fail(f"invalid JSON payload: {exc}")
        if not isinstance(payload, dict):
            fail("JSON payload must be an object")
        return payload

    payload: dict[str, Any] = {
        "project": args.project,
        "feature": args.feature,
        "summary": args.summary,
        "model": args.model,
        "source": args.source,
    }
    for field in INT_FIELDS:
        value = getattr(args, field)
        if value is not None:
            payload[field] = value
    if args.metadata_json:
        try:
            metadata = json.loads(args.metadata_json)
        except json.JSONDecodeError as exc:
            fail(f"invalid metadata JSON: {exc}")
        if not isinstance(metadata, dict):
            fail("metadata must be a JSON object")
        payload["metadata"] = metadata
    return {key: value for key, value in payload.items() if value is not None}


def clean_snapshot(payload: dict[str, Any]) -> dict[str, Any]:
    summary = str(payload.get("summary", "")).strip()
    if not summary:
        fail("summary is required")

    snapshot: dict[str, Any] = {
        "timestamp": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "project": str(payload.get("project") or Path.cwd().name),
        "source": str(payload.get("source") or "manual"),
        "summary": summary,
    }

    if payload.get("model"):
        snapshot["model"] = str(payload["model"])
    if payload.get("feature"):
        snapshot["feature"] = str(payload["feature"])

    for field in INT_FIELDS:
        value = payload.get(field)
        if value is None:
            continue
        try:
            parsed = int(value)
        except (TypeError, ValueError):
            fail(f"{field} must be an integer")
        if parsed < 0:
            fail(f"{field} must be non-negative")
        snapshot[field] = parsed

    if "total_tokens" not in snapshot:
        prompt_tokens = snapshot.get("prompt_tokens")
        completion_tokens = snapshot.get("completion_tokens")
        if isinstance(prompt_tokens, int) and isinstance(completion_tokens, int):
            snapshot["total_tokens"] = prompt_tokens + completion_tokens

    metadata = payload.get("metadata")
    if metadata is not None:
        if not isinstance(metadata, dict):
            fail("metadata must be an object")
        snapshot["metadata"] = metadata

    return snapshot


def append_snapshot(snapshot: dict[str, Any], history_path: Path) -> None:
    history_path.parent.mkdir(parents=True, exist_ok=True)
    with history_path.open("a", encoding="utf-8") as history:
        history.write(json.dumps(snapshot, sort_keys=True, separators=(",", ":")))
        history.write("\n")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--json", help="Snapshot JSON object. If omitted, stdin is used when piped.")
    parser.add_argument("--history", default=os.environ.get("TOKEN_TRACKER_HISTORY"), help="Override history JSONL path.")
    parser.add_argument("--project")
    parser.add_argument("--feature")
    parser.add_argument("--summary")
    parser.add_argument("--model")
    parser.add_argument("--source", default="manual")
    parser.add_argument("--prompt-tokens", dest="prompt_tokens", type=int)
    parser.add_argument("--completion-tokens", dest="completion_tokens", type=int)
    parser.add_argument("--total-tokens", dest="total_tokens", type=int)
    parser.add_argument("--metadata-json")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    payload = load_payload(args)
    snapshot = clean_snapshot(payload)
    history_path = Path(args.history).expanduser() if args.history else DEFAULT_HISTORY
    append_snapshot(snapshot, history_path)
    print(json.dumps({"saved": str(history_path), "snapshot": snapshot}, sort_keys=True))


if __name__ == "__main__":
    main()

File: scripts/save-token-usage.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const os = require("os");
const path = require("path");

const DEFAULT_HISTORY = path.join(os.homedir(), ".cursor", "token-tracker", "history.jsonl");
const INT_FIELDS = ["prompt_tokens", "completion_tokens", "total_tokens"];

function fail(message) {
  console.error(`token-tracker: ${message}`);
  process.exit(2);
}

function expandHome(p) {
  if (!p) return p;
  if (p.startsWith("~/")) return path.join(os.homedir(), p.slice(2));
  return p;
}

function parseArgs(argv) {
  const args = {
    json: null,
    history: process.env.TOKEN_TRACKER_HISTORY || null,
    project: null,
    feature: null,
    summary: null,
    model: null,
    source: "manual",
    prompt_tokens: null,
    completion_tokens: null,
    total_tokens: null,
    metadata_json: null,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    const next = () => {
      i += 1;
      return argv[i];
    };
    if (a === "--json") args.json = next();
    else if (a === "--history") args.history = next();
    else if (a === "--project") args.project = next();
    else if (a === "--feature") args.feature = next();
    else if (a === "--summary") args.summary = next();
    else if (a === "--model") args.model = next();
    else if (a === "--source") args.source = next();
    else if (a === "--prompt-tokens") args.prompt_tokens = Number(next());
    else if (a === "--completion-tokens") args.completion_tokens = Number(next());
    else if (a === "--total-tokens") args.total_tokens = Number(next());
    else if (a === "--metadata-json") args.metadata_json = next();
    else fail(`unknown argument: ${a}`);
  }
  return args;
}

function readStdinSync() {
  try {
    return fs.readFileSync(0, "utf8");
  } catch {
    return "";
  }
}

function loadPayload(args) {
  let raw = args.json;
  if (raw == null && !process.stdin.isTTY) {
    raw = readStdinSync();
  }
  if (raw) {
    let payload;
    try {
      payload = JSON.parse(raw);
    } catch (err) {
      fail(`invalid JSON payload: ${err.message}`);
    }
    if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
      fail("JSON payload must be an object");
    }
    return payload;
  }

  const payload = {
    project: args.project,
    feature: args.feature,
    summary: args.summary,
    model: args.model,
    source: args.source,
  };
  for (const field of INT_FIELDS) {
    if (args[field] != null && !Number.isNaN(args[field])) payload[field] = args[field];
  }
  if (args.metadata_json) {
    let metadata;
    try {
      metadata = JSON.parse(args.metadata_json);
    } catch (err) {
      fail(`invalid metadata JSON: ${err.message}`);
    }
    if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
      fail("metadata must be a JSON object");
    }
    payload.metadata = metadata;
  }
  return Object.fromEntries(Object.entries(payload).filter(([, v]) => v != null));
}

function cleanSnapshot(payload) {
  const summary = String(payload.summary || "").trim();
  if (!summary) fail("summary is required");

  const snapshot = {
    timestamp: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    project: String(payload.project || path.basename(process.cwd())),
    source: String(payload.source || "manual"),
    summary,
  };
  if (payload.model) snapshot.model = String(payload.model);
  if (payload.feature) snapshot.feature = String(payload.feature);

  for (const field of INT_FIELDS) {
    if (payload[field] == null) continue;
    const parsed = Number(payload[field]);
    if (!Number.isInteger(parsed)) fail(`${field} must be an integer`);
    if (parsed < 0) fail(`${field} must be non-negative`);
    snapshot[field] = parsed;
  }

  if (snapshot.total_tokens == null) {
    if (Number.isInteger(snapshot.prompt_tokens) && Number.isInteger(snapshot.completion_tokens)) {
      snapshot.total_tokens = snapshot.prompt_tokens + snapshot.completion_tokens;
    }
  }

  if (payload.metadata != null) {
    if (typeof payload.metadata !== "object" || Array.isArray(payload.metadata)) {
      fail("metadata must be an object");
    }
    snapshot.metadata = payload.metadata;
  }
  return snapshot;
}

function appendSnapshot(snapshot, historyPath) {
  fs.mkdirSync(path.dirname(historyPath), { recursive: true });
  fs.appendFileSync(historyPath, `${JSON.stringify(snapshot)}\n`, "utf8");
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const payload = loadPayload(args);
  const snapshot = cleanSnapshot(payload);
  const historyPath = expandHome(args.history) || DEFAULT_HISTORY;
  appendSnapshot(snapshot, historyPath);
  console.log(JSON.stringify({ saved: historyPath, snapshot }));
}

if (require.main === module) main();

module.exports = { cleanSnapshot, loadPayload, parseArgs };
